compiler/rustc_trait_selection/src/traits/coherence.rs

```diff
@@ -353,49 +353,76 @@ fn negative_impl<'cx, 'tcx>(
     })
 }
 
+/// Try to prove that a negative impl exist for the given obligation and their super predicates.
+#[instrument(level = "debug", skip(selcx))]
 fn negative_impl_exists<'cx, 'tcx>(
     selcx: &SelectionContext<'cx, 'tcx>,
     param_env: ty::ParamEnv<'tcx>,
     region_context: DefId,
     o: &PredicateObligation<'tcx>,
 ) -> bool {
     let infcx = &selcx.infcx().fork();
+
+    if resolve_negative_obligation(infcx, param_env, region_context, o) {
+        return true;
+    }
+
+    // Try to prove a negative obligation exist for super predicates
+    // FIXME we should implement an `assemble_neg_candidates` fn
+    for o in util::elaborate_predicates(infcx.tcx, iter::once(o.predicate)) {
+        if resolve_negative_obligation(infcx, param_env, region_context, &o) {
+            return true;
+        }
+    }
+
+    false
+}
+
+#[instrument(level = "debug", skip(infcx))]
+fn resolve_negative_obligation<'cx, 'tcx>(
+    infcx: &InferCtxt<'cx, 'tcx>,
+    param_env: ty::ParamEnv<'tcx>,
+    region_context: DefId,
+    o: &PredicateObligation<'tcx>,
+) -> bool {
     let tcx = infcx.tcx;
-    o.flip_polarity(tcx)
-        .map(|o| {
-            let mut fulfillment_cx = FulfillmentContext::new();
-            fulfillment_cx.register_predicate_obligation(infcx, o);
-
-            let errors = fulfillment_cx.select_all_or_error(infcx);
-            if !errors.is_empty() {
-                return false;
-            }
 
-            let mut outlives_env = OutlivesEnvironment::new(param_env);
-            // FIXME -- add "assumed to be well formed" types into the `outlives_env`
+    if let Some(o) = o.flip_polarity(tcx) {
```

Comment:
Nit: let else seems appropriate here.

Reply:
:+1:, there's something that I don't like that much about let else, so I end not using it much. Anyway, going to make the change :).